lib/guardian.ex

```diff
@@ -189,6 +189,51 @@ defmodule Guardian do
     end
   end
 
+
+  @doc """
+  Exchange a token with type 'refresh' for a token with type 'token'
+  with the same claims but a shorter expiry time.
+
+  The token with typ 'refresh' will be revoked after the exhange
+
+  """
+  @spec exchange(String.t, String.t, String.t) :: {:ok, String.t, Map} |
+                              {:error, atom} |
+                              {:error, String.t}
+
+  def exchange(old_jwt, from_typ, to_typ) do
+    case decode_and_verify(old_jwt) do
+      {:ok, found_claims} -> do_exchange(from_typ, to_typ, found_claims)
+      {:error, reason} -> {:error, reason}
+    end
+  end
+
+  defp correct_typ?(claims, typ) when is_binary(typ) do
+    Map.get(claims, "typ") === typ
+  end
+
+  defp correct_typ?(claims, typ_list) when is_list(typ_list) do
+    typ = Map.get(claims, "typ")
+    typ_list |> Enum.any?(&(&1 === typ))
+  end
+
+  defp correct_typ?(_claims, _typ) do
+    false
+  end
+
+  defp do_exchange(from_typ, to_typ, claims) do
+    if correct_typ?(claims, from_typ) do
+      {:ok, resource} = Guardian.serializer.from_token(claims["sub"])
+      case encode_and_sign(resource, to_typ, %{}) do
```

Comment:
I think we want to drop some of the claims here and use the remaining claims. If we just use an empty map we'll lose any custom claims that were embedded in the 'refresh' token.